File: interop/test1.pl
% -*- mode: Prolog -*-

% Test of test1.proto, test1.py
% Assumes it is called from test1.pl

:- module(test1, [test1_main/0]).

:- initialization(test1_main, main).

% :- use_module(library(protobufs)). % DO NOT SUBMIT
:- use_module('../protobufs').

:- set_prolog_flag(optimise_debug, false). % ensure assertion/1 is executed

protobufs:enum(Key, Value) :-
    nth0(Value,
         ['E1',
          'Enum2',
          'AnotherEnum'],
         Key).

template1(Template, Vars) :-
    Template = protobuf([
                         double(1,     V_double),
                         float(2,      V_float),
                         unsigned(3,   V_int32),
                         unsigned(4,   V_int64),
                         unsigned(5,   V_uint32),
                         unsigned(6,   V_uint64),
                         integer(7,    V_sint32),
                         integer(8,    V_sint64),
                         integer32(9,  V_fixed32),
                         integer64(10, V_fixed64),
                         integer32(11, V_sfixed32),
                         integer64(12, V_sfixed64),
                         boolean(13,   V_bool),
                         string(14,    V_string),
                         codes(15,     V_bytes),
                         enum(16,      enum(V_enum))
                        ]),
    Vars = [                           V_double,
                                       V_float,
                                       V_int32,
                                       V_int64,
                                       V_uint32,
                                       V_uint64,
                                       V_sint32,
                                       V_sint64,
                                       V_fixed32,
                                       V_fixed64,
                                       V_sfixed32,
                                       V_sfixed64,
                                       V_bool,
                                       V_string,
                                       V_bytes,
                                       V_enum].


test1_main :-
    template1(Template, Vars),
    Vars = [                           V_double,
                                       V_float,
                                       V_int32,
                                       V_int64,
                                       V_uint32,
                                       V_uint64,
                                       V_sint32,
                                       V_sint64,
                                       V_fixed32,
                                       V_fixed64,
                                       V_sfixed32,
                                       V_sfixed64,
                                       V_bool,
                                       V_string,
                                       V_bytes,
                                       V_enum],
    read_file_to_codes('scalars1a_from_python.wire', Wirestream, [encoding(octet),type(binary)]),
    format(user_error, 'Wirestream=~q~n', [Wirestream]),
    protobuf_message(Template, Wirestream),
    print_term(Template, [output(user_error)]), nl,
    assertion(V_double   == 1.0),
    assertion(V_float    == 2.0),
    assertion(V_int32    == 3),
    assertion(V_int64    == 4),
    assertion(V_uint32   == 5),
    assertion(V_uint64   == 6),
    assertion(V_sint32   == 7),
    assertion(V_sint64   == 8),
    assertion(V_fixed32  == 9),
    assertion(V_fixed64  == 10),
    assertion(V_sfixed32 == 11),
    assertion(V_sfixed64 == 12),
    assertion(V_bool     == false),
    assertion(V_string   == "écran 網目錦蛇"),
    assertion(V_bytes    == [0xc3, 0x28]),
    assertion(V_enum     ==  'E1'),
    protobuf_message(Template, Wirestream2),
    open('scalars1a_from_prolog.wire', write, Stream, [encoding(octet),type(binary)]),
    format(Stream, '~s', [Wirestream2]),
    close(Stream).

File: interop/test1.py
# Test of test1.proto, test1.pl
# Assumes that swipl is on the execution path

import contextlib
import os
import shutil
import subprocess
import unittest
from test1_pb2 import Scalars1, Scalars2, Enum

SWIPL = shutil.which('swipl')

scalars1a = Scalars1(
    v_double   = 1.0,
    v_float    = 2.0,
    v_int32    = 3,
    v_int64    = 4,
    v_uint32   = 5,
    v_uint64   = 6,
    v_sint32   = 7,
    v_sint64   = 8,
    v_fixed32  = 9,
    v_fixed64  = 10,
    v_sfixed32 = 11,
    v_sfixed64 = 12,
    v_bool     = False,
    v_string   = 'écran 網目錦蛇',
    v_bytes    = b'\xc3\x28',  # See https://stackoverflow.com/questions/1301402/example-invalid-utf8-string
    v_enum     = Enum.E1,
)

# Test negative values - unsigned are caught by the Python runtime,
# so they get the values "100+..."
scalars1b = Scalars1(
    v_double   = -1.0,
    v_float    = -2.0,
    v_int32    = -3,
    v_int64    = -4,
    v_uint32   = 100+5,
    v_uint64   = 100+6,
    v_sint32   = -7,
    v_sint64   = -8,
    v_fixed32  = 100+9,
    v_fixed64  = 100+10,
    v_sfixed32 = -11,
    v_sfixed64 = -12,
    v_bool     = True,
    v_string   = 'écran 網目錦蛇',
    v_bytes    = b'\xf0\x28\x8c\x28',  # See https://stackoverflow.com/questions/1301402/example-invalid-utf8-string
    v_enum     = Enum.AnotherEnum,
)

class Test1(unittest.TestCase):

    def test1a(self):
        with contextlib.suppress(FileNotFoundError):
            os.remove('scalars1a_from_python.wire')
            os.remove('scalars1a_from_prolog.wire')
        with open('scalars1a_from_python.wire', 'wb') as f:
            f.write(scalars1a.SerializeToString())
        subprocess.run([SWIPL, 'test1.pl'], check=True)
        m2 = Scalars1()
        with open('scalars1a_from_prolog.wire', 'rb') as f:
            m2.ParseFromString(f.read())
        self.assertEqual(m2.v_double,    1.0)
        self.assertEqual(m2.v_float,     2.0)
        self.assertEqual(m2.v_int32,     3)
        self.assertEqual(m2.v_int64,     4)
        self.assertEqual(m2.v_uint32,    5)
        self.assertEqual(m2.v_uint64,    6)
        self.assertEqual(m2.v_sint32,    7)
        self.assertEqual(m2.v_sint64,    8)
        self.assertEqual(m2.v_fixed32,   9)
        self.assertEqual(m2.v_fixed64,   10)
        self.assertEqual(m2.v_sfixed32,  11)
        self.assertEqual(m2.v_sfixed64,  12)
        self.assertEqual(m2.v_bool,      False)
        self.assertEqual(m2.v_string,    'écran 網目錦蛇')
        self.assertEqual(m2.v_bytes,     b'\xc3\x28')
        self.assertEqual(m2.v_enum,      Enum.E1)
        with contextlib.suppress(FileNotFoundError):
            os.remove('scalars1a_from_python.wire')
            os.remove('scalars1a_from_prolog.wire')


if __name__ == '__main__':
    unittest.main()

